xds/internal/balancer/edsbalancer/eds.go

```diff
@@ -103,340 +99,253 @@ func (bb) ParseConfig(c json.RawMessage) (serviceconfig.LoadBalancingConfig, err
-		// only apply to the new EDS response. But this is applied to the RPCs
-		// before the new EDS response. To fully fix this, the EDS balancer
-		// needs to do a graceful switch to another EDS implementation.
-		//
-		// This is OK for now, because we don't actually expect edsServiceName
-		// to change. Fix this (a bigger change) will happen later.
-		b.loadWrapper.UpdateClusterAndService(b.clusterName, b.edsServiceName)
-	}
+	b.logger.Infof("Watch update from xds-client %p, content: %+v", b.xdsClient, pretty.ToJSON(update.eds))
+	b.edsRespReceived = true
+	b.edsResp = update.eds
 
-	// Restart load reporting when the loadReportServer name has changed.
-	if !equalStringPointers(b.loadReportServer, config.LrsLoadReportingServerName) {
-		loadStore := b.startLoadReport(config.LrsLoadReportingServerName)
-		b.loadWrapper.UpdateLoadStore(loadStore)
+	// A new EDS update triggers new child configs (e.g. different priorities
+	// for the priority balancer), and new addresses (the endpoints come from
+	// the EDS response).
+	if err := b.updateChildConfig(); err != nil {
+		b.logger.Warningf("failed to update child policy's balancer config: %v", err)
 	}
-
-	return nil
 }
 
-// startEndpointsWatch starts the EDS watch.
+// updateChildConfig builds a balancer config from eb's cached eds resp and
+// service config, and sends that to the child balancer. Note that it also
+// generates the addresses, because the endpoints come from the EDS resp.
 //
-// This usually means load report needs to be restarted, but this function does
-// NOT do that. Caller needs to call startLoadReport separately.
-func (b *edsBalancer) startEndpointsWatch() {
-	if b.cancelEndpointsWatch != nil {
-		b.cancelEndpointsWatch()
+// If child balancer doesn't already exist, one will be created.
+func (b *edsBalancer) updateChildConfig() error {
+	// Child was build when the first EDS resp was received, so we just build
+	// the config and addresses.
+	if b.child == nil {
+		b.child = newChildBalancer(b.priorityBuilder, b.cc, b.bOpts)
 	}
-	edsToWatch := b.edsToWatch
-	cancelEDSWatch := b.xdsClient.WatchEndpoints(edsToWatch, func(update xdsclient.EndpointsUpdate, err error) {
-		b.logger.Infof("Watch update from xds-client %p, content: %+v", b.xdsClient, pretty.ToJSON(update))
-		b.handleEDSUpdate(update, err)
-	})
-	b.logger.Infof("Watch started on resource name %v with xds-client %p", edsToWatch, b.xdsClient)
-	b.cancelEndpointsWatch = func() {
-		cancelEDSWatch()
-		b.logger.Infof("Watch cancelled on resource name %v with xds-client %p", edsToWatch, b.xdsClient)
-	}
-}
 
-func (b *edsBalancer) cancelWatch() {
-	b.loadReportServer = nil
-	if b.cancelLoadReport != nil {
-		b.cancelLoadReport()
-		b.cancelLoadReport = nil
+	childCfgBytes, addrs, err := buildPriorityConfigJSON(b.edsResp, b.config)
+	if err != nil {
+		err := fmt.Errorf("failed to build priority balancer config: %v", err)
+		b.logger.Warningf("%v", err)
```

Comment:
I think we should avoid logging errors in this function to prevent the same error showing up multiple times in the log.

All call sites of this func are of the form:
```golang
	if err := b.updateChildConfig(); err != nil {
		b.logger.Warningf("failed to update child policy's balancer config: %v", err)
	}
```